Ensure empty body response for 1xx/204/304 per RFC 9112 sec 6.3

bdraco opened the pull request:
<!-- Thank you for your contribution! -->

## What do these changes do?

`Transfer-Encoding` and `Content-Length` will be removed from the server response when sending a 1xx (Informational), 204 (No Content), or 304 (Not Modified) status since no body is expected for these status codes

- `Content-Length` forbidden on 1xx/204 per https://datatracker.ietf.org/doc/html/rfc7230#section-3.3.2 and discouraged on 304 per https://datatracker.ietf.org/doc/html/rfc7232#section-4.1

We need to ensure `Content-Length` is kept for `HEAD` to ensure keep-alive works as expected and is allowed per https://datatracker.ietf.org/doc/html/rfc2616#section-14.13

- `Transfer-Encoding` removed per  https://datatracker.ietf.org/doc/html/rfc9112#section-6.1 
> any recipient on the response chain (including the origin server) can remove transfer codings when they are not needed.

`aiohttp` would incorrectly send an body of `0\r\n\r\n` when `chunked` was set for `Transfer-Encoding` with the above status code. 

This change ensures `aiohttp` does not send these invalid bodies.

## Are there changes in behavior for the user?

The `Transfer-Encoding` and `Content-Length` headers will be removed when sending a 1xx (Informational), 204 (No Content), or 304 (Not Modified) except for `HEAD` responses since these status since no body is expected for these status codes.

An invalid body of `0\r\n\r\n` will no longer be sent with these status codes.

## Related issue number

<!-- Are there any issues opened that will be resolved by merging this change? -->

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [ ] Documentation reflects the changes
- [ ] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names.
- [x] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` for example (588.bugfix)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."


TODO: 

- [x] figure out why client closes connection on HEAD requests with content length
- [x] figure out why client functional tests fail with AIOHTTP_NO_EXTENSIONS=1 `test_keepalive_after_empty_body_status` and `test_keepalive_after_empty_body_status_stream_response` -- They do not fail anymore after the fixed in https://github.com/aio-libs/aiohttp/pull/7755
- [x] timeline
- [x] more functional testing

Dreamsorcerer commented on aiohttp/web_response.py:
```diff
@@ -364,7 +364,25 @@ async def _prepare_headers(self) -> None:
+            # regardless of the header fields present in the message,
+            # and thus cannot contain a message body or trailer section.
+            #
+            if hdrs.TRANSFER_ENCODING in headers:
+                del headers[hdrs.TRANSFER_ENCODING]
+            if hdrs.CONTENT_LENGTH in headers:
+                del headers[hdrs.CONTENT_LENGTH]
```
OK, running through the spec and what code we have here, I think we'll need to rework this a bit.

At line 407, it correctly removes the Content-Length header if needed already (though could do with checking the full range of 1xx codes).

At line 391, I think we just want to make enable_chunking() conditional on the status code. Then pretty much the same as line 407, delete the Transfer-Encoding header for the given status codes (maybe this part can be moved out a level and deduplicated).

This is also supposed to be done for a CONNECT request, regardless of status code. Not sure if we have any code for that (or if the method is available at this point in the code).

We should probably also make sure that any body provided by the user for these status codes results in an exception, if we don't do this already.

Dreamsorcerer commented on tests/test_web_response.py:
```diff
@@ -624,6 +624,23 @@ async def write_headers(status_line, headers):
+        assert hdrs.CONTENT_LENGTH not in headers
+        assert hdrs.TRANSFER_ENCODING not in headers
+
+    writer.write_headers.side_effect = write_headers
```
I feel like we should be able to check this on the resp object normally?

But, either way, this check shouldn't be done for 304.

bdraco replied:
> I feel like we should be able to check this on the resp object normally?
Adjusted in 8f9f7527c455863e254fc7e3c7938f84dedee278 Kept the mock since it needs to be an async mock

> But, either way, this check shouldn't be done for 304.

https://github.com/nodejs/llhttp/pull/256#issuecomment-1782909852
> 
> > Transfer-Encoding MAY be sent in a response to a HEAD request or in a 304 (Not Modified) response ([Section 15.4.5](https://www.rfc-editor.org/rfc/rfc9110#section-15.4.5) of [[HTTP](https://datatracker.ietf.org/doc/html/rfc9110)]) to a GET request, neither of which includes a message body, to indicate that the origin server would have applied a transfer coding to the message body if the request had been an unconditional GET. This indication is not required, however, because any recipient on the response chain (including the origin server) can remove transfer codings when they are not needed.

> It is allowed to send transfer-encoding header field in 304 response, but it must be ignored because the message body must be empty.

I think we should strip it out because it will confuse the older clients if we don't

Dreamsorcerer replied:
Hmm, I guess it's a problem if we have clients that trip up on such a message. But, ideally, it could be used to give additional information about the response... I think the Content-Length header is probably more useful than the Transfer-Encoding header though.

Dreamsorcerer replied:
> Adjusted in [8f9f752](https://github.com/aio-libs/aiohttp/commit/8f9f7527c455863e254fc7e3c7938f84dedee278) Kept the mock since it needs to be an async mock

Would be better if you can successfully create the mock with `mock.create_autospec(Writer, spec_set=True, instance=True)`

bdraco replied:
b8d5991dea9f1754ae8f5d304692aecdd0e987c6

Dreamsorcerer commented:
Thanks for digging into this, I didn't realise from the original issue that it was a response from aiohttp that was causing the problem.

Dreamsorcerer commented on aiohttp/web_response.py:
```diff
@@ -364,7 +364,25 @@ async def _prepare_headers(self) -> None:
+            # regardless of the header fields present in the message,
+            # and thus cannot contain a message body or trailer section.
+            #
+            if hdrs.TRANSFER_ENCODING in headers:
+                del headers[hdrs.TRANSFER_ENCODING]
+            if hdrs.CONTENT_LENGTH in headers:
+                del headers[hdrs.CONTENT_LENGTH]
```
A quick test suggests we don't do the last 2 points currently.

The body is probably a particular issue. If we have a keep-alive connection and we send a body, a conforming client should stop before reading the body, and will then try to interpret the body as a new HTTP message.

bdraco replied:
TODO:

- [ ] At line 407, it correctly removes the Content-Length header if needed already (though could do with checking the full range of 1xx codes).

- [ ] At line 391, I think we just want to make enable_chunking() conditional on the status code. Then pretty much the same as line 407, delete the Transfer-Encoding header for the given status codes (maybe this part can be moved out a level and deduplicated).

- [ ] This is also supposed to be done for a CONNECT request, regardless of status code. Not sure if we have any code for that (or if the method is available at this point in the code).

- [ ] We should probably also make sure that any body provided by the user for these status codes results in an exception, if we don't do this already.

bdraco replied:
>  We should probably also make sure that any body provided by the user for these status codes results in an exception, if we don't do this already.

That seems like it might be better in another PR.

Dreamsorcerer replied:
> > We should probably also make sure that any body provided by the user for these status codes results in an exception, if we don't do this already.
> 
> That seems like it might be better in another PR.

Up to you, but the title of the PR says that it ensures there is an empty body...

bdraco replied:
It will be a bit more refactoring but I can do it here as well

bdraco replied:
I'm not sure we can do that as these tests fail when I do

```
FAILED tests/test_client_functional.py::test_HTTP_304_WITH_BODY[pyloop] - aiohttp.client_exceptions.ServerDisconnectedError: Server disconnected
FAILED tests/test_client_functional.py::test_HTTP_200_OK_METHOD[pyloop] - aiohttp.client_exceptions.ServerDisconnectedError: Server disconnected
FAILED tests/test_client_functional.py::test_HTTP_200_OK_METHOD_connector[pyloop] - aiohttp.client_exceptions.ServerDisconnectedError: Server disconnected
FAILED tests/test_client_functional.py::test_HTTP_302_REDIRECT_HEAD[pyloop] - aiohttp.client_exceptions.ServerDisconnectedError: Server disconnected
FAILED tests/test_client_functional.py::test_shortcuts[pyloop] - aiohttp.client_exceptions.ServerDisconnectedError: Server disconnected
```

Dreamsorcerer replied:
Ah, OK, looking at that first test, it does ensure we end up with an empty body, even if a non-empty body is passed in. So, I guess we can leave it like that... Could maybe just tweak one of your tests (if it's not already done) to include a body as input, and assert there is no body as output (to ensure we cover all the status codes).

bdraco replied:
I think its covered by the keep alive tests but I added an explicit one as well

f9e831a0bb1248537852be21b97894e3d21285ce

bdraco commented:
Probably could use a test to check the raw body from a 304 with chunked pass through as well for the proxy case so it doesn't accidentally add the bad body

bdraco commented:
Need to figure out how I regressed

tests/test_client_functional.py::test_bad_payload_chunked_encoding[pyloop]

bdraco commented:
So it looks like if `writer.enable_chunking()` isn't called which is now the case for `HEAD` it won't keep the connection alive...

need to figure that out later though as out of time

bdraco commented:
Which also means we need keep-alive tests for 304/204/1xx

Dreamsorcerer commented on aiohttp/web_response.py:
```diff
@@ -364,32 +364,52 @@ async def _prepare_headers(self) -> None:
+        # https://datatracker.ietf.org/doc/html/rfc9112#section-6.3
+        # status code: 204, 304, 1xx
+        # method: HEAD
+        request_indicates_empty_body = (
```
Should we be doing this on a CONNECT 2xx response as well?

Dreamsorcerer replied:
I think from my reading, Transfer-Encoding is forbidden, and Content-Length will be the same as 304 (so, if we're removing it for 304 we should remove it here too).

bdraco replied:
I think that works but I don't have a good test case for this

04e6cfaf80c862ed7c805e61ffb60b88e6ad3d9f

Dreamsorcerer commented on aiohttp/web_response.py:
```diff
@@ -364,32 +364,52 @@ async def _prepare_headers(self) -> None:
             # HTTP 1.1: https://tools.ietf.org/html/rfc7230#section-3.3.2
             # HTTP 1.0: https://tools.ietf.org/html/rfc1945#section-10.4
-            elif version >= HttpVersion11 and self.status in (100, 101, 102, 103, 204):
+            elif version >= HttpVersion11 and request_indicates_empty_body:
                 del headers[hdrs.CONTENT_LENGTH]
+                # remove transfer codings when they are not needed
+                # per https://datatracker.ietf.org/doc/html/rfc9112#section-6.1
+                if hdrs.TRANSFER_ENCODING in headers:
+                    del headers[hdrs.TRANSFER_ENCODING]
```
Do you reckon we can just move this out a level and remove the duplicate check at lines 386-389?

```suggestion

        if version >= HttpVersion11 and request_indicates_empty_body:
            # remove transfer codings when they are not needed
            # per https://datatracker.ietf.org/doc/html/rfc9112#section-6.1
            if hdrs.CONTENT_LENGTH in headers:
                del headers[hdrs.CONTENT_LENGTH]
            if hdrs.TRANSFER_ENCODING in headers:
                del headers[hdrs.TRANSFER_ENCODING]
```

bdraco replied:
I think so.  Once I reconcile the other issue I mentioned below I'll do some testing with the suggested change

bdraco replied:
done in 5b557b983ed07c6ebec2c0dfbcd688b8d0ae6c8b

bdraco commented:
There is still another issue here because in Response we also force content length "0" when content length is missing and the body is empty. This is probably for clients that don't handle implicit empty body well.  Will need to reconcile that before addressing review comments

Dreamsorcerer commented:
> There is still another issue here because in Response we also force content length "0" when content length is missing and the body is empty. This is probably for clients that don't handle implicit empty body well.

I think that is needed for the client to know where the message ends. The only exception where it would be acceptable to have no content-length header is when the body must be empty:
https://www.rfc-editor.org/rfc/rfc9110#section-8.6-5

> Also the rfc for http 1.0 says content length 0 is invalid

It says "Any Content-Length greater than or _equal_ to zero is a valid value."

bdraco commented:
> It says "Any Content-Length greater than or **equal** to zero is a valid value."

Right on, I mentally dropped the equal when I read that

bdraco commented on aiohttp/web_response.py:
```diff
@@ -370,27 +380,36 @@ async def _prepare_headers(self) -> None:
+        # HTTP 1.1: https://tools.ietf.org/html/rfc7230#section-3.3.2
+        # HTTP 1.0: https://tools.ietf.org/html/rfc1945#section-10.4
+        if self._must_be_empty_body:
+            # If the content-length is not set to "0" we will
+            # prematurely close a keep-alive connection that could
+            # have been reused. This matches the behavior of Response
+            # which would also set the content-length to "0" in _start
+            headers[hdrs.CONTENT_LENGTH] = "0"
```
If I delete this head instead of setting it to zero the `test_keepalive_after_head_requests_success` test fails.

The test wasn't actually mocking `HEAD` before so it was actually returning an error

bdraco replied:
https://tools.ietf.org/html/rfc7230#section-3.3.2
> A server MAY send a Content-Length header field in a response to a
   HEAD request ([Section 4.3.2 of [RFC7231]](https://datatracker.ietf.org/doc/html/rfc7231#section-4.3.2)); a server MUST NOT send
   Content-Length in such a response unless its field-value equals the
   decimal number of octets that would have been sent in the payload
   body of a response if the same request had used the GET method.


So it looks like we have to delete this

Dreamsorcerer replied:
If you think it improves compatibility with keep-alive clients, it can stay there when it's not a HEAD request. It can't be done for a HEAD request though, as the client may be using the HEAD request to figure out how big the response will be.

Dreamsorcerer replied:
Ah, actually, 304 is the same. Should be left if present, but shouldn't be set to 0.

It appears that 1xx, 204 are actually forbidden to include the header at all:
https://datatracker.ietf.org/doc/html/rfc9110#section-8.6-8

So, yeah, that will need to go.

bdraco commented:
test_keepalive_after_head_requests_success is still failing but I think everything else will pass now

bdraco commented:
Its failing because the waiter is being cancelled in RequestHandler.start which breaks out of the loop

bdraco commented:
Its because connection_lost if called. I think the client is dropping the connection on the head response so there might be a bug reading head responses w/keep alive in the client

bdraco commented:
Out of time to dig for the client side issue so will have to pick this up later

bdraco commented:
The Content-Length entity-header field indicates the size of the entity-body, in decimal number of OCTETs, sent to the recipient or, in the case of the HEAD method, the size of the entity-body that would have been sent had the request been a GET.



So for `HEAD` we should still send `Content-Length`

bdraco commented:
For 204/304 we need to ensure we don't send content-length.. looks like this has tripped up other projects as well https://github.com/heroku/cowboyku/issues/9

bdraco commented:
The tests are failing because of the differences between the c and py parsers which are fixed in https://github.com/aio-libs/aiohttp/pull/7755

bdraco commented:
That's all the time I have for this tonight.

Lots of functional testing tomorrow/monday left to do

bdraco commented:
I've looked it over and read everything in the RFCs again so I think its good to start doing more functional testing with.

I can't think of any more tests to add at this point, but happy to add some more cases if there is something that stands out

Dreamsorcerer commented:
> So for `HEAD` we should still send `Content-Length`

If the user has set it, yes (or we have some other way of determining the size of the GET payload).

Dreamsorcerer commented on aiohttp/helpers.py:
```diff
@@ -1076,3 +1081,13 @@ def status_code_must_be_empty_body(code: int) -> bool:
+    """
+    # https://www.rfc-editor.org/rfc/rfc9110.html#section-8.6-7
+    # https://www.rfc-editor.org/rfc/rfc9110.html#section-8.6-8
+    return code in (204, 304) or 100 <= code < 200 or method == hdrs.METH_CONNECT
```
For CONNECT requests, it only applies if the response is 2xx, so I think there's a small mistake here.

Also, the wording for HEAD and 304 are basically identical, so I'm not 100% sure we should include 1 here and not the other. I think we really just need to ensure that any automatic Content-Length header is excluded in both cases.

Dreamsorcerer replied:
Just realised the 2xx CONNECT is also an issue with the previously merged functions.

bdraco replied:
I will prepare another PR to fix that

bdraco replied:
https://github.com/aio-libs/aiohttp/pull/7772

I will adjust the logic in this PR shortly

bdraco replied:
logic adjusted in 5630ca593341df97d5551d3dd175163afcbc05d6

bdraco replied:
Tests adjusted in 79d55e28b248c48b90b53eef0d0239fdc1d17728

Dreamsorcerer commented on aiohttp/web_fileresponse.py:
```diff
@@ -266,7 +266,7 @@ async def prepare(self, request: "BaseRequest") -> Optional[AbstractStreamWriter
 
         # If we are sending 0 bytes calling sendfile() will throw a ValueError
-        if count == 0 or request.method == hdrs.METH_HEAD or self.status in [204, 304]:
+        if count == 0 or must_be_empty_body(request.method, self.status):
```
If the problem here is sending 0 bytes, then that'd include both 304 and HEAD, right?

Dreamsorcerer replied:
Ignore that one, mixing up the methods.

Dreamsorcerer commented on aiohttp/web_response.py:
```diff
@@ -370,26 +374,33 @@ async def _prepare_headers(self) -> None:
+                request.method, self.status
+            ):
+                del headers[hdrs.CONTENT_LENGTH]
+            # remove transfer codings when they are not needed
+            # per https://datatracker.ietf.org/doc/html/rfc9112#section-6.1
+            if version >= HttpVersion11 and hdrs.TRANSFER_ENCODING in headers:
```
```suggestion
            # https://datatracker.ietf.org/doc/html/rfc9112#section-6.1-10
            # https://datatracker.ietf.org/doc/html/rfc9112#section-6.1-13
            if hdrs.TRANSFER_ENCODING in headers:
```
The second link says not to send Transfer-Encoding at all for HTTP/1.0.

Dreamsorcerer commented on aiohttp/web_response.py:
```diff
@@ -662,7 +673,12 @@ async def write_eof(self, data: bytes = b"") -> None:
 
     async def _start(self, request: "BaseRequest") -> AbstractStreamWriter:
-        if not self._chunked and hdrs.CONTENT_LENGTH not in self._headers:
+        if self._must_be_empty_body and should_remove_content_length(
```
Isn't the first check redundant, as a superset of the second?

Dreamsorcerer commented on tests/test_client_functional.py:
```diff
@@ -82,6 +83,77 @@ async def on_reuseconn(session, ctx, params):
     await resp2.read()
 
     assert 1 == cnt_conn_reuse
+    await client.close()
```
This shouldn't be needed, our fixtures handle closing the client.

bdraco replied:
Will validate and remove

bdraco replied:
removed in a7dc41ce44c4c735c6a98234ab2c072c95e7d33a

bdraco commented on aiohttp/web_response.py:
```diff
@@ -662,7 +673,12 @@ async def write_eof(self, data: bytes = b"") -> None:
+        ):
+            if hdrs.CONTENT_LENGTH in self._headers:
+                del self._headers[hdrs.CONTENT_LENGTH]
+        elif not self._chunked and hdrs.CONTENT_LENGTH not in self._headers:
```
I think its ok to run on HEAD

> A server MAY send a Content-Length header field in a response to a HEAD request ([Section 9.3.2](https://www.rfc-editor.org/rfc/rfc9110.html#HEAD)); a server MUST NOT send Content-Length in such a response unless its field value equals the decimal number of octets that would have been sent in the content of a response if the same request had used the GET method.

bdraco replied:
https://www.rfc-editor.org/rfc/rfc9110.html#section-8.6-7

Dreamsorcerer replied:
Hmm, I think if the HEAD request was handled automatically by aiohttp, then it will be fine, as it just treats it as a GET request and will have included the entire body of the response (which will then be stripped later). We should ensure there is a test that verifies the HEAD request doesn't receive a 0 Content-Length.

But, if a user manually handles a HEAD request, then this behaviour is incorrect, as they may return an empty body for the HEAD request, but not for a GET request. I don't know if we have an easy way to tell if this is from a user handler or an automatic aiohttp one...

bdraco replied:
> Hmm, I think if the HEAD request was handled automatically by aiohttp, then it will be fine, as it just treats it as a GET request and will have included the entire body of the response (which will then be stripped later). We should ensure there is a test that verifies the HEAD request doesn't receive a 0 Content-Length for this case.

I think this is covered by `test_head_response_keeps_content_length_of_original_body`
> 
> But, if a user manually handles a HEAD request, then this behaviour is incorrect, as they may return an empty body for the HEAD request, but not for a GET request. I don't know if we have an easy way to tell if this is from a user handler or an automatic aiohttp one...

Need to think about how to handle this case

Dreamsorcerer replied:
> I think this is covered by `test_head_response_keeps_content_length_of_original_body`

Yes, and I can see in test_web_functional.py we have a HEAD test there: test_head_returns_empty_body
Would be ideal to just add 1 more assert to that test for the Content-Length too.

Dreamsorcerer replied:
> I think its ok for 304 as well

I think as per the above discussion, we can't do this for 304, as it will always be a user-supplied response. So, it should skip that step in all cases of 304, I believe.

Dreamsorcerer replied:
The user can obviously set Content-Length themselves if they want to include that information.

bdraco replied:
adjusted `test_head_returns_empty_body` 02f22fa3626fd1c1e80003845902d1c38548837f

bdraco replied:
> I think as per the above discussion, we can't do this for 304, as it will always be a user-supplied response. So, it should skip that step in all cases of 304, I believe.

I should probably do that in `_prepare_headers` instead... will adjust

bdraco replied:
oh, its already done there

bdraco replied:
`test_rm_content_length_1xx_204_304_responses` covers the 304

Dreamsorcerer replied:
> oh, its already done there

That's because the helper function is still checking 304, right? The spec treats it the same as HEAD requests. So, basically don't delete the header if the user set it explicitly, but also don't add the header automatically if they didn't.

I guess the HEAD request case, we can go with the check of whether the body is set, as you've done (but, that does still leave a small chance for issues if the user passes a body by mistake, but that's probably the best we'll get for now).

Dreamsorcerer replied:
I had a quick look to see if we could get back to the handlers/urldispatcher (as we could then check if the same handler is registered for GET/HEAD), but looks like there's no path to it. So, will stick with what you've done.

Dreamsorcerer commented on aiohttp/helpers.py:
```diff
@@ -1065,14 +1065,37 @@ def parse_http_date(date_str: Optional[str]) -> Optional[datetime.datetime]:
+    return (
+        status_code_must_be_empty_body(code)
+        or method_must_be_empty_body(method)
+        or (code == 200 and method.upper() == hdrs.METH_CONNECT)
```
I believe it was anything in the 2xx range, not just 200 itself.

bdraco replied:
Will hunt for it in the RFC and adjust as needed as well as the other PR

bdraco replied:
> A server MUST NOT send a Content-Length header field in any response with a status code of [1xx (Informational)](https://www.rfc-editor.org/rfc/rfc9110.html#status.1xx) or [204 (No Content)](https://www.rfc-editor.org/rfc/rfc9110.html#status.204). A server MUST NOT send a Content-Length header field in any [2xx (Successful)](https://www.rfc-editor.org/rfc/rfc9110.html#status.2xx) response to a CONNECT request ([Section 9.3.6](https://www.rfc-editor.org/rfc/rfc9110.html#CONNECT)).

Yes. 2xx 👍

bdraco replied:
adjusted in 286a1084fd75344bda821116d0109e152dcaf6d5

Dreamsorcerer commented on aiohttp/helpers.py:
```diff
@@ -1076,3 +1085,17 @@ def status_code_must_be_empty_body(code: int) -> bool:
+    # https://www.rfc-editor.org/rfc/rfc9110.html#section-8.6-7
+    # https://www.rfc-editor.org/rfc/rfc9110.html#section-8.6-8
+    return (
+        code in {204, 304}
```
304 should still be dropped from here, right?
If the user sets the header, we should keep it.

bdraco replied:
I'm leaning towards it should still be removed since `Content-Length` is discouraged (SHOULD NOT) for 304 since its not one of `Cache-Control`, `Content-Location`, `Date`, `ETag`, `Expires`, and `Vary`.

If you feel strongly otherwise, I can adjust the behavior to keep it in, I adjust all the tests as well, and add a test for keeping it.

https://datatracker.ietf.org/doc/html/rfc7232#section-4.1

> The server generating a 304 response MUST generate any of the
> following header fields that would have been sent in a 200 (OK)
> response to the same request: Cache-Control, Content-Location, Date,
> ETag, Expires, and Vary.

> Since the goal of a 304 response is to minimize information transfer
> when the recipient already has one or more cached representations, a
> sender SHOULD NOT generate representation metadata other than the
> above listed fields unless said metadata exists for the purpose of
> guiding cache updates (e.g., Last-Modified might be useful if the
> response does not have an ETag field).

Dreamsorcerer replied:
I'll come back tomorrow to give this a last check, but that sounds reasonable. They should really have linked to that paragraph from the other section that makes it sound as if it should be treated the same as HEAD.

bdraco commented:
I'll get the backport done shortly

bdraco commented:
Thanks @Dreamsorcerer
